lib/ts/recipe/anomalydetection/types.ts

```diff
@@ -0,0 +1,113 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import OverrideableBuilder from "supertokens-js-override";
+import type { BaseRequest, BaseResponse } from "../../framework";
+import NormalisedURLDomain from "../../normalisedURLDomain";
+import NormalisedURLPath from "../../normalisedURLPath";
+
+export type TypeInput = {
+    basePath?: string;
+    domain?: string;
+    mode?: "BLOCK_ACCESS" | "REPORT_ONLY";
+    override?: {
+        functions?: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis?: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeNormalisedInput = {
+    domain: NormalisedURLDomain;
+    basePath: NormalisedURLPath;
+    mode: "BLOCK_ACCESS" | "REPORT_ONLY";
+    override: {
+        functions: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type APIOptions = {
+    recipeImplementation: RecipeInterface;
+    config: TypeNormalisedInput;
+    recipeId: string;
+    isInServerlessEnv: boolean;
+    req: BaseRequest;
+    res: BaseResponse;
+};
+
+type CheckAnomalyBaseRequest = {
+    action: "SIGN_UP" | "SIGN_IN" | "REFRESH_TOKEN";
+} & (
+    | {
+          status: "OK";
+          userId: string;
+          tenantId: string;
+      }
+    | {
+          status:
+              | "LINKING_TO_SESSION_USER_FAILED"
+              | "WRONG_CREDENTIALS_ERROR"
+              | "SIGN_IN_NOT_ALLOWED"
+              | "EMAIL_ALREADY_EXISTS_ERROR";
+          userId?: string;
+          tenantId?: string;
+      }
+    | {}
+);
+
+export type RecipeInterface = {
+    checkAnomaly: (
+        input: {
+            ipAddress?: string;
+            headers?: DeviceHeaders;
+        } & CheckAnomalyBaseRequest
+    ) => Promise<
+        | {
+              status: "OK";
+          }
+        | {
+              status: "ANOMALY_DETECTED";
+          }
+    >;
+    checkAnomalyWithRequest: (
+        input: {
+            request: BaseRequest;
+        } & CheckAnomalyBaseRequest
+    ) => Promise<
+        | {
+              status: "OK";
+          }
+        | {
+              status: "ANOMALY_DETECTED";
+          }
+    >;
+    getIpAddressFromRequest: (input: { request: BaseRequest }) => string | undefined;
+    getDeviceHeadersFromRequest: (input: { request: BaseRequest }) => DeviceHeaders | undefined;
+};
+
+export type APIInterface = {};
+
+export type DeviceHeaders = {
+    accept?: string;
+    language?: string;
+    encoding?: string;
+    userAgent?: string;
```

Comment:
maybe its better to just get all the headers and pass them as key value map to your service. Then your service can just pick which ones it cares about?

Reply:
I though about that but could end up with a huge payload without real reason. For the moment that is all I need.

Reply:
Not passing all may cause issues with improving the microservice. Cause lets say the microservice needs something else later on, but then that header is not a part of this object. Now we will have to ask users to also update the backend sdk in their integration.

Reply:
ok